src/main/java/com/mongodb/spark/sql/connector/read/MongoPartitionReader.java

```diff
@@ -0,0 +1,119 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.mongodb.spark.sql.connector.read;
+
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.read.InputPartition;
+import org.apache.spark.sql.connector.read.PartitionReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.bson.BsonDocument;
+
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoCursor;
+
+import com.mongodb.spark.sql.connector.assertions.Assertions;
+import com.mongodb.spark.sql.connector.config.ReadConfig;
+import com.mongodb.spark.sql.connector.schema.BsonDocumentToRowConverter;
+
+/**
+ * A partition reader returned by {@link MongoPartitionReaderFactory#createReader(InputPartition)}.
+ * It's responsible for outputting data for a RDD partition.
+ */
+public class MongoPartitionReader implements PartitionReader<InternalRow> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoPartitionReader.class);
+  private final MongoInputPartition partition;
+  private final BsonDocumentToRowConverter bsonDocumentToRowConverter;
+  private final ReadConfig readConfig;
+
+  private boolean closed = false;
+  private MongoClient mongoClient;
+  private MongoCursor<BsonDocument> mongoCursor;
+
+  /**
+   * Construct a new instance
+   *
+   * @param partition the partition
+   * @param bsonDocumentToRowConverter the converter from {@link BsonDocument} to {@link
+   *     InternalRow}
+   * @param readConfig the read configuration for reading from the partition
+   */
+  public MongoPartitionReader(
+      final MongoInputPartition partition,
+      final BsonDocumentToRowConverter bsonDocumentToRowConverter,
+      final ReadConfig readConfig) {
+    this.partition = partition;
+    this.bsonDocumentToRowConverter = bsonDocumentToRowConverter;
+    this.readConfig = readConfig;
+    LOGGER.debug(
+        "Creating partition reader for: PartitionId: {} with Schema: {}",
+        partition.getPartitionId(),
+        bsonDocumentToRowConverter.getSchema());
+  }
+
+  /** Proceed to next record, returns false if there is no more records. */
+  @Override
+  public boolean next() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call next() on a closed PartitionReader.");
+    return getCursor().hasNext();
+  }
+
+  /** Return the current record. This method should return same value until `next` is called. */
+  @Override
+  public InternalRow get() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call get() on a closed PartitionReader.");
+    return bsonDocumentToRowConverter.toInternalRow(getCursor().next());
```

Comment:
I think the current implementation has two problems:

1. `PartitionReader.get` [_"should return same value until `next` is called"_](https://spark.apache.org/docs/3.1.2/api/java/org/apache/spark/sql/connector/read/PartitionReader.html#get--). This requirement is violated.
2. The fact that `PartitionReader.next` is declared as `throws IOException`, but `PartitionReader.get` is not, implies that `PartitionReader.next` must do the IO, while `PartitionReader.get` should only return the data read by `PartitionReader.next`.

Reply:
Good catch - done

Reply:
`get` now returns the same value until `next` is called, however, the `next` method does not necessary ["_Proceed to next record_"](https://spark.apache.org/docs/3.1.2/api/java/org/apache/spark/sql/connector/read/PartitionReader.html#next--). If we are going to base the implementation of `MongoPartitionReader.get` and `MongoPartitionReader.next` on the documentation of the methods `MongoCursor.hasNext`/`MongoCursor.next`<sup>1</sup>, then in order to satisfy the specification of the method `PartitionReader.next` we have to

- call `MongoCursor.next` in `MongoPartitionReader.next` if `MongoCursor.hasNext` produces `true`;
- store the result of `MongoCursor.next` for later use by `MongoPartitionReader.get`.

---
<sup>1</sup> As we should, unless there is a good reasons to break encapsulation and start depending on the undocumented behavior.

Reply:
Updated.